text/0000-import-trait-methods.md

```diff
@@ -0,0 +1,253 @@
+}
+
+impl S {
+    fn new() -> S {
+        S {
+            a: default()
+        }
+    }
+}
+```
+
+You can also use this with trait methods that take a `self` argument:
+
+```rust
+use num_traits::float::Float::{sin, cos}
+
+fn eulers_formula(theta: f64) -> (f64, f64) {
+    (cos(theta), sin(theta))
+}
+```
+
+Importing a method from a trait does not import the trait. If you want to call `self` methods on a trait or `impl` it, then you can import the trait as well as methods in the trait:
+
+```rust
+mod a {
+    pub trait A {
+        fn new() -> Self;
+        fn do_something(&self);
+    }
+}
+
+mod b {
+    use super::a::A::{self, new}
+
+    struct B();
+
+    impl A for B {
+        fn new() -> Self {
+            B()
+        }
+
+        fn do_something(&self) {
+        }
+    }
+
+    fn f() {
+        let b: B = new();
+        b.do_something();
+    }
+}
+```
+
+Trait methods can also be renamed when they are imported using the usual `as` syntax:
+```rust
+use Default::default as gimme
+
+struct S {
+    a: HashMap<i32, i32>,
+}
+
+impl S {
+    fn new() -> S {
+        S {
+            a: gimme()
+        }
+    }
+}
+```
+
+You can also import a parent trait method from a sub-trait:
+
+```rust
+use num_traits::float::Float::zero;
+
+fn main() {
+    let x : f64 = zero();
+    println!("{}",x);
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+When 
+
+```rust
+use Trait::method as m;
+```
+occurs, we first find the supertrait in which `method` occurs. If it occurs in multiple supertraits, or in the trait and in a supertrait, then error.
+A new item `m` is made available in the function namespace of the current module. Any attempts to call this item are treated calling the (super-)trait method explicitly qualified. As always, the `as` qualifier is optional, in which case the name of the new item is identical with the name of the method in the trait. 
+
+In other words, the example:
+
+```rust
+use Default::default;
+
+struct S {
+    a: HashMap<i32, i32>,
+}
+
+impl S {
+    fn new() -> S {
+        S {
+            a: default()
+        }
+    }
+}
+```
+desugars to 
+```rust
+struct S {
+    a: HashMap<i32, i32>,
+}
+
+impl S {
+    fn new() -> S {
+        S {
+            a: Default::default()
+        }
+    }
+}
+```
+And a call
+```rust
+use Trait::method as m;
+m(x, y, z);
+```
+desugars to
+```rust
+Trait::method(x, y, z);
+```
+
+Additionally, the syntax
+```rust
+use some_module::Trait::self;
+```
+is sugar for
+```rust
+use some_module::Trait;
+```
+
+Finally, given traits 
+```rust
+trait Super {
+    fn f();
+}
+
+trait Sub : Super {
+}
+```
+the usage
+```rust
+use module::Sub::f;
+f();
+```
+desugars to
+```rust
+use module::Sub::f;
+Super::f();
+```
+**not** `Sub::f();` as that desugaring will cause compiler errors, see https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=51bef9ba69ce1fc20248e987bf106bd4.
+
+# Drawbacks
```

Comment:
another drawback is that this makes already complicated name resolution rules even more complicated, which have been a frequent cause of compiler bugs in the past and are particularly difficult to maintain already. They are also hard to understand for users and equally difficult for alternative implementations like rust-analyzer.